daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/HexBinaryCodeGenerator.scala

```diff
@@ -60,76 +62,70 @@ trait HexBinaryCodeGenerator extends BinaryValueCodeGenerator {
     else
-      s"""    $field.array = NULL;
-         |    $field.lengthInBytes = 0;
-         |    $field.dynamic = false;""".stripMargin
-    val initSelfStatement = if (specifiedLength > 0)
-      s"    memset($fieldArray, 0x77, sizeof($fieldArray));"
-    else
-      ""
+      s"""$indent1$indent2    $field.array = NULL;
+         |$indent1$indent2    $field.lengthInBytes = 0;
+         |$indent1$indent2    $field.dynamic = false;""".stripMargin
     val parseStatement =
-      s"""    parse_hexBinary(&$field, pstate);
-         |    if (pstate->error) return;""".stripMargin
+      s"""$indent1$indent2    parse_hexBinary(&$field, pstate);
+         |$indent1$indent2    if (pstate->error) return;""".stripMargin
     val unparseStatement =
-      s"""    unparse_hexBinary($field, ustate);
-         |    if (ustate->error) return;""".stripMargin
-    cgState.addSimpleTypeStatements(initERDStatement, initSelfStatement, parseStatement, unparseStatement)
+      s"""$indent1$indent2    unparse_hexBinary($field, ustate);
+         |$indent1$indent2    if (ustate->error) return;""".stripMargin
+    cgState.addSimpleTypeStatements(initERDStatement, parseStatement, unparseStatement)
   }
 
   // Generate C code to validate a hexBinary element against its fixed value
   private def hexBinaryValidateFixed(e: ElementBase, deref: String, cgState: CodeGeneratorState): Unit = {
+    val indent1 = if (cgState.hasChoice) "    " else ""
+    val indent2 = if (deref.nonEmpty) "    " else ""
     val localName = e.namedQName.local
     val field = s"instance->$localName$deref"
-    val i = if (deref.length > 2) deref.substring(1, deref.length - 1) else ""
-    val fixed = s"${localName}_fixed$i"
+    val fixed = s"${localName}_fixed"
     val array = e.fixedValueAsString.grouped(2).mkString("0x", ", 0x", "")
 
     val initERDStatement = ""
-    val initSelfStatement = ""
     val parseStatement =
-      s"""    uint8_t $fixed[] = {$array};
-         |    parse_validate_fixed(memcmp($field.array, $fixed, sizeof($fixed)) == 0, "$localName", pstate);
-         |    if (pstate->error) return;""".stripMargin
+      s"""$indent1$indent2    uint8_t $fixed[] = {$array};
+         |$indent1$indent2    parse_validate_fixed(memcmp($field.array, $fixed, sizeof($fixed)) == 0, "$localName", pstate);
+         |$indent1$indent2    if (pstate->error) return;""".stripMargin
     val unparseStatement =
-      s"""    uint8_t $fixed[] = {$array};
-         |    unparse_validate_fixed(memcmp($field.array, $fixed, sizeof($fixed)) == 0, "$localName", ustate);
-         |    if (ustate->error) return;""".stripMargin
-    cgState.addSimpleTypeStatements(initERDStatement, initSelfStatement, parseStatement, unparseStatement)
+      s"""$indent1$indent2    uint8_t $fixed[] = {$array};
+         |$indent1$indent2    unparse_validate_fixed(memcmp($field.array, $fixed, sizeof($fixed)) == 0, "$localName", ustate);
+         |$indent1$indent2    if (ustate->error) return;""".stripMargin
```

Comment:
I wonder if there's a better way to handle all this indentation? Perhaps we could create a custom function similar to stripMargin but strips margin + indents to some level.

```scala
val indentLevel = 1 + Seq(cgState.hasChoice, deref.nonEmpty).count(_ == true)

s"""|uint8_t $fixed[] = {$array};
    |unparse_validate_fixed(memcmp($field.array, $fixed, sizeof($fixed)) == 0, "$localName", ustate);
    |if (ustate->error) return;
    |""".indent(indentLevel)
```

Not sure I like the indentLevel logic one-liner, but it something along those lines could be a bit cleaner and make it easier to see the real code and reduce duplication.

Reply:
I like the idea, but how about calling the function `stripMargin(indent: Int)` if Scala allows us to overload the original `stripMargin` function?  Keeping the original name and allowing any number of spaces makes the function more general and powerful.

```scala
val indent = 2 + 2 * Seq(cgState.hasChoice, deref.nonEmpty).count(_ == true)

s"""|uint8_t $fixed[] = {$array};
    |unparse_validate_fixed(memcmp($field.array, $fixed, sizeof($fixed)) == 0, "$localName", ustate);
    |if (ustate->error) return;
    |""".stripMargin(indent)
```

Reply:
Sees reasonable to me if Scala has no problem with it. stripMargin does have an overload with a Char parameter. I assume Scala can figure out the differences between an int and a char, but if not a variable on stripMargin seems reasonable.

Reply:
I haven't implemented stripMargin(indent) yet since it doesn't have to be done right now.